cpp/src/copying/gather.cu

```diff
@@ -0,0 +1,380 @@
+ *
+ * If any of the range [source_mask, source_mask + num_source_rows) overlaps
+ * [destination_mask, destination_mask + num_destination_rows), the result is
+ * undefined.
+ *
+ * @tparam T The type of the stencil array
+ * @tparam P The type of the predicate
+ * @param[in] source_mask The mask whose bits will be gathered
+ * @param[in] num_source_rows The number of bits in the source_mask
+ * @param[out] destination_mask The output after gathering the input
+ * @param[in] num_destination_rows The number of bits in the
+ * destination_mask
+ * @param[in] gather_map The map that indicates where elements from the
+ * input will be gathered to in the output. Length must be equal to
+ * `num_destination_rows`.
+ * @param[in] stencil An array of values that will be evaluated by the
+ * predicate. Length must be equal to `num_destination_rows`.
+ * @param[in] pred Unary predicate applied to the stencil values
+ *---------------------------------------------------------------------------**/
+template <typename T, typename P>
+__global__ void gather_bitmask_if_kernel(
+    gdf_valid_type const* const __restrict__ source_mask,
+    gdf_size_type const num_source_rows, gdf_valid_type* const destination_mask,
+    gdf_size_type const num_destination_rows, gdf_index_type const* gather_map,
+    T const* stencil, P pred) {
+  using MaskType = uint32_t;
+  constexpr uint32_t BITS_PER_MASK{sizeof(MaskType) * 8};
+
+  // TODO: Update to use new bit_mask_t
+  MaskType* const __restrict__ destination_mask32 =
+      reinterpret_cast<MaskType*>(destination_mask);
+
+  gdf_index_type destination_row = threadIdx.x + blockIdx.x * blockDim.x;
+
+  auto active_threads =
+      __ballot_sync(0xffffffff, destination_row < num_destination_rows);
+
+  while (destination_row < num_destination_rows) {
+    bool const source_bit_is_valid{
+        gdf_is_valid(source_mask, gather_map[destination_row])};
+
+    // Use ballot to find all valid bits in this warp and create the output
+    // bitmask element
+    MaskType const result_mask{__ballot_sync(
+        active_threads, pred(stencil[destination_row]) && source_bit_is_valid)};
+
+    gdf_index_type const output_element = destination_row / BITS_PER_MASK;
+
+    // Only one thread writes output
+    if (0 == threadIdx.x % warpSize) {
+      destination_mask32[output_element] = result_mask;
+    }
+
+    destination_row += blockDim.x * gridDim.x;
+    active_threads =
+        __ballot_sync(active_threads, destination_row < num_destination_rows);
+  }
+}
+
+/**---------------------------------------------------------------------------*
+ * @brief Gathers the bits of a validity bitmask.
+ *
+ * Gathers the bits from the source bitmask into the destination bitmask
+ * according to a `gather_map` such that bit `i` in `destination_mask` will be
+ * equal to bit `gather_map[i]` from `source_bitmask`.
+ *
+ * Undefined behavior results if any value in `gather_map` is outside the range
+ * [0, num_destination_rows).
```

Comment:
You mean `num_source_row`? Since gather_map is the same size as `num_destination_rows` and contains indices to `source_mask`

Reply:
Ah yes! Thank you for catching that.